SvelteKit and Directus as a CMS (#28)
directus-community/guest-authoring: sveltekit-directus/index.md

	directus.readItems = async (collection, query) => directus.request(readItems(collection, query));
	directus.readItem = async (collection, id, query) => directus.request(readItem(collection, id, query));
	directus.createItem = async (collection, query) => directus.request(createItem(collection, query));
	directus.deleteItem = async (collection, id) => directus.request(deleteItem(collection, id));

	return directus;
}

export default getDirectusInstance;
```

In order to make this work we also need to create a `hooks.server.js` file with the following content in the root
directory. It makes sure that the required headers for fetching JavaScript content are returned by the SvelteKit Server

```js
export async function handle({ event, resolve }) {
	return await resolve(event, {
		filterSerializedResponseHeaders: (key, value) => {
			return key.toLowerCase() === 'content-type';
		},
	});
}
```

::: tip Directus SDK

Theoretically you could also make HTTP requests to your Directus server endpoint directly via SvelteKit's `fetch`
implementation. However the Directus SDK offers some nice [additional features](/guides/sdk/getting-started).

> :::

Author: however we will need to use this for authentication as the directus SDK does not play nice together with sveltekits SSR. but sure it can be removed here. i just think one should in general also explain users what the options are in order to better understand the reasoning behind choosing a path

> :::

Reviewer: With that context I'm happy to leave it

## Setup Global Metadata and Settings

In your Directus project, navigate to Settings -> Data Model and create a new collection called `global`. Under the
Singleton option, select 'Treat as a single object', as this collection will have just a single entry containing global
website metadata.

Create two text input fields - one with the key of `title` and one `description`.

Navigate to the content module and enter the global collection. Collections will generally display a list of items, but
as a singleton, it will launch directly into the one-item form. Enter information in the title and description field and
hit save.

![A form named "Global" has two inputs - a title and a description, each filled with some text.](global-config.webp)

By default, new collections are not accessible to the public. Navigate to Settings -> Roles & Permissions -> Public and
give Read access to the Global collection.

## Prepare SvelteKit to use Directus instance

Create a new file called `+page.js` in the root directory along the `.page.svelte` file. This file's load function will
be responsible to fetch the data on the client and on the server during Server Side Rendering.

> 

Author: added file headings to all code snippets

```js
/** @type {import('./$types').PageLoad} */
import getDirectusInstance from '$lib/directus';
export async function load({ fetch }) {
	const directus = getDirectusInstance(fetch);
	return {
		global: await directus.readItems('global'),
	};
}
```

> Correspondingly we need to modify the `+page.svelte` file to take the data into account and display it on our site.
> SvelteKit magically populates the data property and also updates it should any dependency(variable, url, etc.) change
> which influences the data fetch mechanism in the `+page.js` file. Currently there is no dependency, later you will see
> how this works.

Reviewer: ```suggestion
Modify the `+page.svelte` file to use the new data and display it on our site:
```

Remove half explaining things in advance. It doesn't make sense yet, so save it for when it does.

Author: moved it to a later section and shortened it. would include it however, as it is quite important to understand SSR

```svelte
<script>
	/** @type {import('./$types').PageData} */
	export let data;
</script>

<h1>{data.global.title}</h1>
<p>{data.global.description}</p>
```

Refresh your browser. You should see data from your Directus Global collection in your page.

## Creating Pages With Directus

> Create a new directory called `[slug]`. SvelteKit uses a file based routing mechanism and parameters are always part of
> the directory name, whereas the files within the directory are always either +page.js, +page.js or +page.server.js (which is
> only run on server side only so it is save to expose secrets in this file). The same concept is true for layout files.
> Please refer to the [SvelteKit Documentation](https://kit.svelte.dev/docs/routing) for more information.

Reviewer: ```suggestion
Create a new directory called `[slug]`. SvelteKit uses a file based routing mechanism and parameters are always part of
the directory name, while the files within the directory are always either `+page.js`, `+page.js` or `+page.server.js`.
```

We're not teaching people SvelteKit - just how Directus plays with it.

Public role read access to the new collection. Create a few items in the new collection.

Inside of `[slug]`, create a new file called `+page.js`. This is a dynamic route, so this time we will use the dynamic
`params` object to fetch the correct data. To illustrate how SvelteKit's data loading works you can open a different
page url which will change the `params` object. Evidently this will lead to SvelteKit invalidate the `.page.js` data and
refetch our page data.

```js
/** @type {import('./$types').PageLoad} */
import { error } from '@sveltejs/kit';
import getDirectusInstance from '$lib/directus';
export async function load({ fetch, params }) {
	const directus = getDirectusInstance(fetch);

	try {
		return {
			page: await directus.readItem('pages', params.slug),
		};
	} catch (err) {
		throw error(404, 'Page not found');
	}
}
```

Go to http://localhost:5173/about, replacing `about` with any of your item slugs. Using the Directus JavaScript SDK, the
single item with that slug is retrieved, and the page should show your data. `readItem()` only checks against your
`slug` Primary ID Field.

> _Note that non existing IDs will result in a forbidden error, which we catch and then throw svelte's internal error
> object to respond with a 404. Please also note that
> [`@html` should only be used for trusted content.](https://svelte.dev/docs/special-tags#html)_

Reviewer: ```suggestion
:::warning 404s and Trusted Content

Non-existing IDs will result in a forbidden error, which we catch and then throw svelte's internal error
object to respond with a 404. Additionally,
[`@html` should only be used for trusted content.](https://svelte.dev/docs/special-tags#html)_

:::
```

			<span>{post.publish_date} &bull; {post.author.name}</span>
		</li>
	{/each}
</ul>
```

Visit http://localhost:5173 and you should now see a blog post listing, with latest items first.

![A page with a title of "Blog". On it is a list of three items - each with a title, author, and date. The title is a link.](blog-listing.webp)

### Create Blog Post Listing

Each blog post links to a page that does not yet exist. In the `blog` directory, create a new directory called `[slug]`
with the necessary files as usual:

::: code-group

```js [+page.js]
/** @type {import('./$types').PageLoad} */
import { error } from '@sveltejs/kit';
import getDirectusInstance from '$lib/directus';
export async function load({ fetch, params }) {
	const directus = getDirectusInstance(fetch);

	try {
		return {
			post: await directus.readItem('posts', params.slug, {
				fields: ['*', { '*': ['*'] }],
			}),
		};
	} catch (err) {
		throw error(404, 'Post not found');
	}
}
```

```svelte [+page.svelte]
<script>
	/** @type {import('./$types').PageData} */
	export let data;
</script>

<img src="{import.meta.env.VITE_APIURL}/assets/{data.post.image.filename_disk}?width=600" alt="" />
<h1>{data.post.title}</h1>

> <div>{@html data.post.content}</div>

Reviewer: Would be good to use the image description as the alt here
